system/jlib/jlzw.cpp

```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
     return (size32_t)(in-(const byte *)src);
 }
 
-void appendToBuffer(MemoryBuffer & out, size32_t len, const void * src)
+void compressToBuffer(MemoryBuffer & out, size32_t len, const void * src, CompressionMethod method, const char *options)
 {
-    out.append(false);
-    out.append(len);
-    out.append(len, src);
-}
-
-void compressToBuffer(MemoryBuffer & out, size32_t len, const void * src)
-{
-    unsigned originalLength = out.length();
-    out.append(true);
-    out.append((size32_t)0);
-
-    if (len >= 32)
+    if (method != COMPRESS_METHOD_NONE && len >= 32)
     {
+        ICompressHandler *handler = queryCompressHandler(method);
+        if (!handler)
+        {
+            VStringBuffer s("Unknown compression method %x requested in compressToBuffer", (byte) method);
+            throw makeStringException(0, s.str());
+        }
+        unsigned originalLength = out.length();
+        // For back-compatibility, we always store COMPRESS_METHOD_LZW as 1 as earlier versions stored a boolean here
+        // rather than an enum
+        // This means that compressToBuffer/decompressToBuffer cannot bs used for rowdiff compression - this is not likely to be an issue
+        // Alternative would be a separate enum for compressToBuffer formats, but that seems more likely to cause confusion
+        out.append((byte) (method == COMPRESS_METHOD_LZW ? COMPRESS_METHOD_LZWLEGACY : method));
+        out.append((size32_t)0);
         size32_t newSize = len * 4 / 5; // Copy if compresses less than 80% ...
-        Owned<ICompressor> compressor = createLZWCompressor();
+        Owned<ICompressor> compressor = handler->getCompressor(options);
         void *newData = out.reserve(newSize);
-        compressor->open(newData, newSize);
-        if (compressor->write(src, len)==len)
+        try
```

Comment:
Future: Could use compressToBlock() if supported by the compressor.  There is also a few bytes wasted, which would be more significant for small sizes, but that requires substantial refactoring.

Reply:
I have created a separate jira for this HPCC-30928 (and a branch gch/issue30928).  It has two effects
i) it compresses better due to the daft way the current lz4 compress works.
ii) it is faster.
Before:
```
12:24:58.258158 419929 compressToBuffer 5 size 0 did not compress
12:24:58.258387 419929 compressToBuffer 5 size 102 compressed to 66 in 1985ns
12:24:58.258803 419929 compressToBuffer 5 size 1000 compressed to 53 in 1503ns
```
After:
```
12:25:21.154246 420137 compressToBuffer 5 size 0 did not compress
12:25:21.154313 420137 compressToBuffer 5 size 34 compressed to 32 in 2434ns
12:25:21.154783 420137 compressToBuffer 5 size 1000 compressed to 35 in 1082ns
```
We could remove 4 bytes from the size, but that would be trickier (the format wouldn't be compatible).

With that change, switching to a less repetitive text  (Gen 1 from 0drvb10), LZ4 is better at compressing, but significantly slower:

```
12:37:32.829295 422255 compressToBuffer 2 size 512 compressed to 357 in 5490ns
12:37:32.837024 422255 compressToBuffer 2 size 1000 compressed to 627 in 9808ns
12:37:32.838347 422255 compressToBuffer 5 size 512 compressed to 411 in 1893ns
12:37:32.841931 422255 compressToBuffer 5 size 1000 compressed to 717 in 2885ns
```